pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -571,6 +575,17 @@ protected void handleSend(CommandSend send, ByteBuf headersAndPayload) {
             printSendCommandDebug(send, headersAndPayload);
         }
 
+        // avoid processing non-persist message if reached max concurrent-message limit
+        if (producer.isNonPersistentTopic() && !nonPersistentMessageSemaphore.tryAcquire()) {
+            final long producerId = send.getProducerId();
+            final long sequenceId = send.getSequenceId();
+            service.getTopicOrderedExecutor().submitOrdered(producer.getTopic(), SafeRun.safeRun(() -> {
```

Comment:
actually, we have to write it in ordered thread only else send-ack goes in random order and because of that client will not receive sequenceId in order and it will fail the pending msgs and close connection.
- if msg doesn't throttle here then broker processes it by delivering it to subscribers in ordered thread and after completion that ordered thread sends ack back to client.
- so, all the sending-ack operation has to be in ordered-thread only.